<img src='https://lh3.googleusercontent.com/b8Cmw_ouaEGOrfNOonyOfHYwvZCQDgPwdpkZGyZr3UEpbKb_vLfBHRgJcuZW_GtWmi61ODYvHjACskutfMLkriDyWhVNfwEJrbqYjTFjuH8wsUcWaby-hCYRQErX3SigVZbF0pKjOQ' alt="" width="700" height="250"/>


<br><br>
<p style="font-family: Arial;font-size:3.5em; color:#09bef3; font-style:bold; text-align:center">
LimFinity Institute of Technology
</p>

<p style="font-family: Arial; font-size:2em;color:#09bef3; font-style:normal; text-align:center; font-weight: bold"><br><br>

مشروع التنبؤ بالطقس
<br>
<br><br>
<img src='https://lh3.googleusercontent.com/VEN8wyE_iRsbaqDsSaSybFZkA8vpSD8SYLuRLlpImC1LbYijmJQ_am0MqAZQhgNeiSWESaE_-gW-D-08FnoYhWehpWLnryS-edTEsAXAllV6zZHRsVvsUxgF-1An9gCpIAKugR9rvA' alt="" width="800" height="200"/></a>


<p style="font-family: Arial; font-size:1.75em;color:#09bef3; font-style:bold; text-align:right; direction: rtl; font-weight: bold" ><br>

مصدر البيانات: بيانات الطقس اليومي مجموعة بواسطة جامعة كاليفورنيا سان دييغو
</p>

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

في هذا المشروع سوف نقوم ببناء نظام ذكي يتنبئ بالطقس (الرطوبة) هل الرطوبة عالية ام منخفظة من خلال استخدام تعليم الالة وسوف نستخدم في هذا المشروع التصنيف (Classification) نظام التعليم بالاشراف وسوف نقوم بأستخدام خوارزمية Decision Tree وسوف نتعلم كيف نقوم ببناء نموذج تعليم الالة واساسيات بناء النموذج ثم تجريبه ثم قياس اداءه (الخطوات السبعة التي تم شرحها في الدورة)

<br><br>
البيانات التي سنعمل عليها تحتوي على 1095 سطر كل سطر يمثل بيانات يوم يعني لدينا 1095 يوم (ثلاث سنوات من سبتمبر 2011 الى سبتمبر 2014 من اجل ضمان عدد كافي من البيانات لبناء نموذج) هذه البيانات سجلت بواسطة مستشعرات وحساسات عالية الدقة حيث في كل يوم تم تسجيل بيانات سرعة الرياح ودرجة الحرارة والامطار والرطوبة النسبية وتم تسجيل هذه المعلومات في 11 عمود هذه الاعمدة هي:

</p>

* **number:** (رقم فريد لكل سطر)
* **air_pressure_9am:** (معدل الضغط الجوي في الساعة التاسعة صباحا)
* **air_temp_9am:** (معدل درجة الحرارة في الساعة التاسعة صباحا)
* **air_wind_direction_9am:** (معدل اتجاه الرياح في الساعة التاسعة صباحا)
* **air_wind_speed_9am:** (معدل سرعة الرياح في الساعة التاسعة صباحا)
* ** max_wind_direction_9am:** (اقصى اتجاه للرياح في الساعة التاسعة صباحا)
* **max_wind_speed_9am:** (اقصى سرعة للرياح في الساعة التاسعة صباحا)
* **rain_accumulation_9am:** (كمية الامطار المتراكمة في ال 24 ساعة الماضية للساعة التاسعة صباحا)
* **rain_duration_9am:** (كمية الوقت الذي هطلت فيه امطار في ال 24 ساعة الماضية للساعة التاسعة صباحا)
* **relative_humidity_9am:** (معدل الرطوبة النسبية في الساعة التاسعة صباحا)
* **relative_humidity_3pm:** (معدل الرطوبة النسبية في الساعة الثالثة مساءا)


<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

الهدف من هذا المشروع هو بناء تطبيق يتنبئ بالرطوبة للأيام (هل الرطوبة عالية ام منخفظة) وسوف نفعل ذلك من خلال بناء نموذج تعليم الة يتعلم من بيانات الصباح 9am لتوقع رطوبة المساء 3pm
</p>

<p style="font-family: Arial; font-size:1.75em;color:#09bef3; font-style:bold; text-align:right; direction: rtl; font-weight: bold" ><br>

استيراد المكتبات التي نحتاجها في المشروع:
</p>

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

- سوف نقوم بأستيراد مكتبة Pandas من اجل قراءة البيانات واستكشافها
<br>
- وسوف نقوم بأستيراد train_test_split من مكتبة sklearn من اجل قسم البيانات الى قسمين هما train لتدريب النموذج و test لأختبار النموذج
<br>
- وسوف نقوم بأستيراد الخوارزمية Decision Tree Classifier احدى خوارزميات التصنيف المهمة من اجل بناء نموذج تصنيفي بأستخدامها وسوف نتعرف عليها اكثر في الدورات القادمة
<br>
- وسوف نقوم بأستيراد accuracy_score من اجل فحص دقة النموذج

</p>

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score


<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>
سوف نقوم بقراء البيانات daily_weather.csv والمخزونة في الفولدر datasets ومن الواضح انه صيغة هذا الملف comma-separated-values CSV وسوف نحول هذا الملف الى جدول DataFrame

بأستخدام Pandas وسوف نخزنه في متغير اسمه data
</p>

In [2]:
data = pd.read_csv('datasets/daily_weather.csv')

<p style="font-family: Arial; font-size:1.75em;color:#09bef3; font-style:bold; text-align:right; direction: rtl; font-weight: bold" ><br>

استكشاف البيانات:
</p>


<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

وكالمعتاد اول خطوة يجب القيام بها بداية اي مشروع هو استكشاف البيانات من اجل فهمها جيدا حيث اول شيء سنفعله هو استعراض الاعمدة من خلال المثود columns

</p>

In [3]:
data.columns

Index(['number', 'air_pressure_9am', 'air_temp_9am', 'avg_wind_direction_9am',
       'avg_wind_speed_9am', 'max_wind_direction_9am', 'max_wind_speed_9am',
       'rain_accumulation_9am', 'rain_duration_9am', 'relative_humidity_9am',
       'relative_humidity_3pm'],
      dtype='object')

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

الان لنستكشف اول خمسة سطور من بياناتنا من اجل التعرف على شكل البيانات بداخلها 
</p>

In [4]:
data.head()

number  air_pressure_9am  air_temp_9am  avg_wind_direction_9am  \
0       0        918.060000     74.822000              271.100000   
1       1        917.347688     71.403843              101.935179   
2       2        923.040000     60.638000               51.000000   
3       3        920.502751     70.138895              198.832133   
4       4        921.160000     44.294000              277.800000   

   avg_wind_speed_9am  max_wind_direction_9am  max_wind_speed_9am  \
0            2.080354              295.400000            2.863283   
1            2.443009              140.471548            3.533324   
2           17.067852               63.700000           22.100967   
3            4.337363              211.203341            5.190045   
4            1.856660              136.500000            2.863283   

   rain_accumulation_9am  rain_duration_9am  relative_humidity_9am  \
0                    0.0                0.0              42.420000   
1                    0.0                0.0              24.328697   
2                    0.0               20.0               8.900000   
3                    0.0                0.0              12.189102   
4                    8.9            14730.0              92.410000   

   relative_humidity_3pm  
0              36.160000  
1              19.426597  
2              14.460000  
3              12.742547  
4              76.740000

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

لنتعرف على نوع البيانات في كل عمود ولنتعرف على عدد سطور كل عمود من خلال المثود info</p>

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1095 entries, 0 to 1094
Data columns (total 11 columns):
number                    1095 non-null int64
air_pressure_9am          1092 non-null float64
air_temp_9am              1090 non-null float64
avg_wind_direction_9am    1091 non-null float64
avg_wind_speed_9am        1092 non-null float64
max_wind_direction_9am    1092 non-null float64
max_wind_speed_9am        1091 non-null float64
rain_accumulation_9am     1089 non-null float64
rain_duration_9am         1092 non-null float64
relative_humidity_9am     1095 non-null float64
relative_humidity_3pm     1095 non-null float64
dtypes: float64(10), int64(1)
memory usage: 94.2 KB


<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

من الواضح ان جميع الاعمدة رقمية ومن الواضح انه هناك نقوصات في عدد السطور في الاعمدة ولذلك لنتأكد هل يوجد قيم فارغة في الاعمدة بأستخدام المثود isnull
</p>

In [6]:
data.isnull().any()

number                    False
air_pressure_9am           True
air_temp_9am               True
avg_wind_direction_9am     True
avg_wind_speed_9am         True
max_wind_direction_9am     True
max_wind_speed_9am         True
rain_accumulation_9am      True
rain_duration_9am          True
relative_humidity_9am     False
relative_humidity_3pm     False
dtype: bool

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

نعم هناك قيم فارغة والان لنتأكد من عدد القيم الفارغة في كل عمود حتى نعرف هل عددها كبير جدا يستحق حذف العمود بالكامل ام عددها قليل ولايؤثر على عملنا حتى نحذف السطور التي تحتوي على قيم فارغة فقط
</p>

In [7]:
data.isnull().sum()

number                    0
air_pressure_9am          3
air_temp_9am              5
avg_wind_direction_9am    4
avg_wind_speed_9am        3
max_wind_direction_9am    3
max_wind_speed_9am        4
rain_accumulation_9am     6
rain_duration_9am         3
relative_humidity_9am     0
relative_humidity_3pm     0
dtype: int64

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

من الواضح ان عددها قليل ولاتؤثر على عملنا ولكن لنستعرض السطور التي تحتوي على قيم فارغة قبل حذفها لانه ربما تحتوي على شيء مميز او سبب معين او ربما ليس هناك اي شيء مجرد قيمة فارغة غير مسجلة (كعالم بيانات او عالم ذكاء اصطناعي يجب ان يكون لديك الفضول لمعرفة كل شيء)

</p>

In [8]:
# to show the null values
data[data.isnull().any(axis=1)]

number  air_pressure_9am  air_temp_9am  avg_wind_direction_9am  \
16        16        917.890000           NaN              169.200000   
111      111        915.290000     58.820000              182.600000   
177      177        915.900000           NaN              183.300000   
262      262        923.596607     58.380598               47.737753   
277      277        920.480000     62.600000              194.400000   
334      334        916.230000     75.740000              149.100000   
358      358        917.440000     58.514000               55.100000   
361      361        920.444946     65.801845               49.823346   
381      381        918.480000     66.542000               90.900000   
409      409               NaN     67.853833               65.880616   
517      517        920.570000     53.600000              100.100000   
519      519        916.250000     55.670000              176.400000   
546      546               NaN     42.746000              251.100000   
620      620        921.200000     56.786000              192.300000   
625      625        912.400000     50.774000              171.600000   
656      656        920.830000     66.344000                     NaN   
670      670        910.920000     48.362000              156.500000   
672      672        922.448945     72.863773                     NaN   
705      705        911.900000     59.072000              199.800000   
731      731        922.970166     51.391847               33.810942   
737      737        917.895130     76.804690              104.771020   
788      788        917.923442     73.249717               42.101739   
840      840        918.043767           NaN              181.774042   
848      848        915.250000     37.562000              246.500000   
861      861        919.065408           NaN              172.303728   
869      869               NaN     45.104000              259.000000   
998      998        914.140000     71.240000                     NaN   
1031    1031        922.669195           NaN               47.946284   
1035    1035        919.670000     77.576000              171.800000   
1063    1063        917.300185     65.790001                     NaN   
1066    1066        919.564869     73.726732               68.704694   

      avg_wind_speed_9am  max_wind_direction_9am  max_wind_speed_9am  \
16              2.192201              196.800000            2.930391   
111            15.613841              189.000000                 NaN   
177             4.719943              189.900000            5.346287   
262            10.636273               67.145843           13.671423   
277             2.751436                     NaN            3.869906   
334             2.751436              187.500000            4.183078   
358            10.021491                     NaN           12.705819   
361            21.520177               61.886944           25.549112   
381             3.467257               89.400000            4.406772   
409             4.328594               78.570923            5.216734   
517             4.697574                     NaN            6.285801   
519             6.666081              188.200000                 NaN   
546            12.929513              274.400000           17.604718   
620             9.551734              201.400000           11.005745   
625                  NaN              181.400000            4.831790   
656            15.457255              189.400000           16.486248   
670                  NaN              177.500000           16.128337   
672             3.682370              214.196160            4.849450   
705             1.275056              239.500000            1.834291   
731                  NaN               59.290089           11.111555   
737             1.632705               97.178763                 NaN   
788             4.132698               64.284969            5.345258   
840             0.964376              185.618

<p style="font-family: Arial; font-size:1.75em;color:#09bef3; font-style:bold; text-align:right; direction: rtl; font-weight: bold" ><br>

تنظيف البيانات:
</p>


<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

من اجل بناء نموذج تعليم الة يجب ان لايكون لدينا اي قيم فارغة في البيانات ولذلك علينا معالجة القيم المفقودة قبل بناء النموذج وايضا يجب التخلص من الاعمدة الغير مفيدة وعلى سبيل المثال العمود number يحتوي على رقم السطر يعني ليس له اي تأثير على بيانات الطقس هو موجود فقط للتسلسل ولذلك سوف نتخلص منه لانه الاعمدة الغير مفيدة تؤثر على نتائج النموذج الذي سوف نبنيه وسوف نتعلم في الدورات القادمة كيف نعرف الاعمدة المفيدة والاكثر تأثيرا على النموذج وكيف نعرف على الاعمدة الغير مفيدة من اجل ازالتها
<br>
والان سوف نحذف العمود number لكونه غير مفيد وليس له علاقة بالطقس


</p>

In [9]:
del data['number']


<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

قبل حذف السطور التي تحتوي على القيم الفارغة لنقوم بخزن عدد السطور قبل حذف القيم الفارغة 

</p>

In [10]:
data.shape

(1095, 10)

In [11]:
before_rows = data.shape[0]
print(before_rows)

1095


<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>
الان لنقوم بحذف السطور التي تحتوي على قيم فارغة من خلال dropna وتعني اسقط القيم الفارغة ولنخزن البيانات في المتغير data من جديد ولكن هذه المرة data بدون قيم فارغة
</p>

In [12]:
data = data.dropna()

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>
لنقم بحساب عدد السطور بعد حذف السطور التي تحتوي على قيم فارغة
</p>

In [13]:
after_rows = data.shape[0]
print(after_rows)

1064


<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>
والان لنكتشف كم سطر تم حذفه من اجل تنظيف البيانات من القيم الفارغة

</p>

In [14]:
before_rows - after_rows

31

<p style="font-family: Arial; font-size:1.75em;color:#09bef3; font-style:bold; text-align:right; direction: rtl; font-weight: bold" ><br>

تحضير البيانات لبناء النموذج:
</p>


<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

والان بعد ان استكشفنا البيانات وفهمناها وبعد ان قمنا بتنظيف البيانات سوف نبدأ الان بتحضير البيانات من اجل بناء النموذج مثلا تحضير X, y ومن ثم قسم البيانات الى تدريبية وتجريبية <br>

</p>
<p style="font-family: Arial; font-size:1.5em;color:black; font-style:normal; text-align:right; direction: rtl;font-weight:bold"><br>

انشاء X, y
</p>

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>
كما تعلمنا في الدرس ان X هي اعمدة الخصائص و y هي عمود ال Label

<br>
وقبل البدأ في انشاء X, y لنأخذ نسخة نظيفة من البيانات بأستخدام copy
</p>

In [15]:
clean_data = data.copy()

<p style="font-family: Arial; font-size:1.5em;color:black; font-style:normal; text-align:right; direction: rtl;font-weight:bold"><br>

التحويل الى مهمة تصنيفية وتهيئة عمود الهدف y للنموذج

</p>

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

البيانات التي نعمل عليها تحتوي على عدة اعمدة وكما علمنا سابقا انه الهدف هو التنبؤ بالرطوبة مساءا هل هي عالية ام منخفضة يعني الهدف هو عمود الرطوبة النسبية مساءا relative_humidity_3pm هذا يعني ان هذا العمود هو ال Label ولذلك سيكون هو ال y للنموذج 
<br><br>
لكن قبل خزن هذا العمود في متغير اسمه y هناك مشكلة تواجهنا وهي انه هذا العمود عبارة عن ارقام الرطوبة النسبية ومن اجل بناء نموذج Classification يجب ان يكون عمود ال Label تصنيفي مثلا 0 او 1 او نعم او لا ثم نحولها الى 0 او 1 او غائم صحو ممطر ونحولها الى ارقام 1 او 2 او 3 لانه هذه من شروط النموذج التصنيفي هي ان يكون عمود ال Label مصنف

<br>

لذلك سوف نقوم بتحويل العمود relative_humidity_3pm الى 0 و 1 حيث كل نسبة رطوبة اقل من 25 سنمثلها ب 0 وهي تمثل رطوبة منخفضة وكل نسبة رطوبة اعلى من 25 سنمثلها ب 1 يعني رطوبة عالية

<br>

من اجل القيام بذلك سوف نصنع عمود جديد اسمه الرطوبة العالية high_humidity_label قيمته هي فلتر بسيط يقارن هل عمود الرطوبة النسبية اكبر من 25 ومن ثم سنضرب هذا الفلتر ب 1 حتى اذا كانت قيمة الفلتر True سنحصل على رقم 1 واذا كانت قيمة الفلتر False سنحصل على الرقم 0 وكما في الكود التالي

</p>

In [16]:
clean_data['high_humidity_label'] = (clean_data['relative_humidity_3pm'] > 24.99)*1 
print(clean_data['high_humidity_label'])

0       1
1       0
2       0
3       0
4       1
5       1
6       0
7       1
8       0
9       1
10      1
11      1
12      1
13      1
14      0
15      0
17      0
18      1
19      0
20      0
21      1
22      0
23      1
24      0
25      1
26      1
27      1
28      1
29      1
30      1
       ..
1064    1
1065    1
1067    1
1068    1
1069    1
1070    1
1071    1
1072    0
1073    1
1074    1
1075    0
1076    0
1077    1
1078    0
1079    1
1080    0
1081    0
1082    1
1083    1
1084    1
1085    1
1086    1
1087    1
1088    1
1089    1
1090    1
1091    1
1092    1
1093    1
1094    0
Name: high_humidity_label, Length: 1064, dtype: int32



<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

الان لنخزن نسخة من هذا العمود في y حيث y ستكون هي ال Label للنموذج وتسمى ايضا الهدف Target

</p>

In [17]:
y = clean_data[['high_humidity_label']].copy()

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>
ولتوضيح ماقمنا به سوف نقارن بين قيمة عمود الرطوبة النسبية في الثالثة مساءا قبل التحويل الى عمود مصنف وبعد تحويله الى عمود مصنف وخزنه في y ولاحظ ايضا ان كل رقم اكبر من 25 اصبح 1 وكل رقم اصغر من 25 اصبح 0
</p>

In [18]:
clean_data['relative_humidity_3pm'].head()

0    36.160000
1    19.426597
2    14.460000
3    12.742547
4    76.740000
Name: relative_humidity_3pm, dtype: float64

In [19]:
y.head()

high_humidity_label
0                    1
1                    0
2                    0
3                    0
4                    1

<p style="font-family: Arial; font-size:1.5em;color:black; font-style:normal; text-align:right; direction: rtl;font-weight:bold"><br>

تهيئة الخصائص X للنموذج
</p>


<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

ببساطة مانريده هو التنبؤ بالرطوبة في المساء pm من خلال استخدام بيانات الطقس في الصباح am ولذلك اعمدة الخصائص التي سندرب النموذج عليها ستكون جميع اعمدة بيانات الطقس في الصباح وسوف نخزن اسماء اعمدة الخصائص التي نريدها في متغير اسمه morning_features ومن ثم سندخله على البيانات من اجل اختيار فقط هذه الاعمدة من البيانات ومن ثم سنخزنها في X
</p>

In [20]:
morning_features = ['air_pressure_9am','air_temp_9am','avg_wind_direction_9am','avg_wind_speed_9am',
        'max_wind_direction_9am','max_wind_speed_9am','rain_accumulation_9am',
        'rain_duration_9am']
morning_features

['air_pressure_9am',
 'air_temp_9am',
 'avg_wind_direction_9am',
 'avg_wind_speed_9am',
 'max_wind_direction_9am',
 'max_wind_speed_9am',
 'rain_accumulation_9am',
 'rain_duration_9am']

In [21]:
X = clean_data[morning_features].copy()
X.head()

air_pressure_9am  air_temp_9am  avg_wind_direction_9am  avg_wind_speed_9am  \
0        918.060000     74.822000              271.100000            2.080354   
1        917.347688     71.403843              101.935179            2.443009   
2        923.040000     60.638000               51.000000           17.067852   
3        920.502751     70.138895              198.832133            4.337363   
4        921.160000     44.294000              277.800000            1.856660   

   max_wind_direction_9am  max_wind_speed_9am  rain_accumulation_9am  \
0              295.400000            2.863283                    0.0   
1              140.471548            3.533324                    0.0   
2               63.700000           22.100967                    0.0   
3              211.203341            5.190045                    0.0   
4              136.500000            2.863283                    8.9   

   rain_duration_9am  
0                0.0  
1                0.0  
2               20.0  
3                0.0  
4            14730.0

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

وللتوضيح مرة اخرى لنستعرض اعمدة X التي سندرب عليها النموذج ثم لسنتعرض عمود الهدف y والذي هو ال Label للنموذج</p>

In [22]:
X.columns

Index(['air_pressure_9am', 'air_temp_9am', 'avg_wind_direction_9am',
       'avg_wind_speed_9am', 'max_wind_direction_9am', 'max_wind_speed_9am',
       'rain_accumulation_9am', 'rain_duration_9am'],
      dtype='object')

In [23]:
y.columns

Index(['high_humidity_label'], dtype='object')

<p style="font-family: Arial; font-size:1.75em;color:#09bef3; font-style:bold; text-align:right; direction: rtl; font-weight: bold" ><br>

قسم البيانات الى تدريبية وتجريبية (Train and Test split):

</p>

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

كما تعلمنا في الدرس انه من اجل بناء نموذج نحتاج في المرحلة الاولى الى تدريبه على بيانات تدريبية Training Data مع اختيار خوارزمية للنموذج والتعديل على معلمات هذه الخوارزمية لحين الوصول الى نموذج ممتاز والصورة التالية توضح مرحلة بناء وتدريب النموذج

</p>

<img class="aligncenter" src="https://lh3.googleusercontent.com/8R1_tb-AuhB8Mr1gZZHilIjt7MJ1iZhSRJgoiqFpekSobxydnZw5BwEU9tl1yI-ItGUS6g0oVDi9AFaZYPoruZAl01rV0ZY4B0HrNrpavwIV-iEYPuppEOQo_pHmx3ds3GjqFQJl4g" width="800" />

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

وبعد بناء النموذج نحتاج الى تجريبه ومن اجل تجريب النموذج نحتاج الى بيانات لم يتعرف عليها النموذج مسبقا ولم يراها ابدا لانه الهدف هو بناء نموذج يؤدي اداءا جيدا على البيانات التدريبية وعلى البيانات التجريبية لانه في المستقبل سيعمل هذا النموذج على بيانات جديدة لم يراها ويجب ان يكون اداءه جيد عليها
<br><br>
ولذلك نحتاج الى قسم بياناتنا الى تدريبية وتجريبية حتى ندرب النموذج ونجربه ايضا وعادة تكون نسبة البيانات التدريبية تقريبا 80% من البيانات الكلية او اقل وربما اكثر والباقي يكون للبيانات التجريبية وهذه النسب تعتمد على البيانات فهي غير ثابتة بأمكانك تغييرها

<br>
والصورة التالية توضح مرحلة الاختبار بأستخدام البيانات التجريبية وبأستخدام النموذج الذي سوف ننشأه
</p>

<img class="aligncenter" src="https://lh3.googleusercontent.com/JFOTzibil9lUW8qzW_GxYtZGZVOrIVYYJpDWV0PswA899Jr2RJWXo-77HDQT9T7YROpu-ybTJ0T_1n34SaV3pxFq-moPPYqyG6u58x1K0jP8cqJqbzqt5YHS4kdTpIcfKZxD0VrfXQ" width="800" />

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

من اجل قسم البيانات نحتاج الى استخدام train_test_split من مكتبة sklearn ومن ثم ندخل لها X, y التي صنعناها ومن ثم نحدد حجم البيانات التجريبية مثلا 0.2 وتعني 20% ونضع رقم عشوائي مثلا 123 حتى تكون النتيجة ثابتة في كل مرة نقسم فيها البيانات
<br><br>
هذه الوظيفة سوف تقسم X الى قسمين وسنخزن هذين القسمين في المتغيرين التاليين X_train, X_test
<br>
وسوف تقسم y الى قسمين ايضا وسوف نخزنهما في المتغيرين التاليين y_train, y_test

</p>

In [24]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)

<p style="font-family: Arial; font-size:1.75em;color:#09bef3; font-style:bold; text-align:right; direction: rtl; font-weight: bold" ><br>

بناء النموذج وتدريبه:
</p>

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

بعد ان قسمنا البيانات نستطيع الان بناء النموذج بأستخدام البيانات التدريبية X_train, y_train ولكننا نحتاج ايضا الى اختيار خوارزمية معينة لبناء هذا النموذج وكما ذكرنا في بداية هذا المشروع اننا سوف نستخدم خوارزمية DecisionTreeClassifier في هذا المشروع وسوف نتعرف اكثر على تفاصيل الخوارزميات في الدورات القادمة ولكن ببساطة هذه الخوارزمية تستخدم في المهام التصنيفية مثل المهمة التي لدينا وهي التنبؤ بالرطوبة عالية ام منخفضة ومن اهم المعلمات لها هو عمق او عدد العقد max_leaf_nodes وهنا للتجربة سنستخدم 10 وسوف نتعلم في الدورات القادمة كيف نختار افضل قيم للمعلمات من اجل الحصول على نموذج دقيق جدا
<br>
<br>
وكما شرحنا في الدرس ان الخوارزميات في مكتبة sklearn عبارة عن بايثون كلاس يجب ان نخزنها في متغير من اجل تحويلها الى كائن حتى نستطيع استخدام وظائفها ولذلك سوف نخزنها في متغير سوف نسميه humidity_classifier
<br>
ثم سنقوم بتدريب النموذج على X_train, y_train بأستخدام الوظيفة fit


</p>




In [25]:
humidity_classifier = DecisionTreeClassifier(max_leaf_nodes=10, random_state=0)
humidity_classifier.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=10,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=0,
            splitter='best')

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

بعد خزن الخوارزمية في humidity_classifier وبعد تدريبها على البيانات التدريبية اصبح الان humidity_classifier هو النموذج (خوارزمية + بيانات تدريبية)
<br>
لنطبع نوع المتغير (النموذج) humidity_classifier من اجل التأكد انه تم تدريبه على الخوارزمية DecisionTreeClassifier وسنجد انه تحول الى نموذج مبني على هذه الخوارزمية ومدرب على البيانات التدريبية
</p>

In [26]:
type(humidity_classifier)

sklearn.tree.tree.DecisionTreeClassifier

<p style="font-family: Arial; font-size:1.75em;color:#09bef3; font-style:bold; text-align:right; direction: rtl; font-weight: bold" ><br>

التنبؤ واختبار النموذج:
</p>

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

الان سوف نستخدم البيانات التجريبية X_test والتي لم يتدرب عليها النموذج ولم يتعرف عليها ابدا وسوف نقوم بالتنبؤ بال Label لهذه البيانات

<br>
مع العلم انه لدينا ال Label للبيانات X_test وهو y_test ولكن الفكرة هي انه سنجعل النموذج يتنبئ بال Label للبيانات التجريبية X_test ومن ثم سنقارن بين ال Label الذي تنبئ به النموذج لهذه البيانات والذي سوف نسميه y_pred وبين ال Label الحقيقي للبيانات التجريبية وهو y_test من اجل قياس دقة النموذج في التنبؤ

<br>
ومن اجل التنبؤ بال Label للبيانات التجريبية X_test نحتاج الى استخدام وظيفة التنبؤ predict وسوف نخزن الناتج في المتغير y_pred
</p>




In [27]:
y_pred = humidity_classifier.predict(X_test)

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

والان لنستعرض اول عشرة Label تنبئ بها النموذج للبيانات التجريبية والتي خزناها في y_pred ولنقارنها بأول عشرة Label حقيقية للبيانات التجريبية
</p>


In [28]:
y_pred[:10]

array([1, 1, 1, 0, 0, 0, 1, 1, 0, 1])

In [29]:
y_test['high_humidity_label'][:10]

712    1
360    0
203    1
845    0
284    0
13     1
364    1
600    0
264    0
278    1
Name: high_humidity_label, dtype: int32

<p style="font-family: Arial; font-size:1.75em;color:#09bef3; font-style:bold; text-align:right; direction: rtl; font-weight: bold" ><br>

قياس دقة نموذج التصنيف:
</p>

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

من الصعب جدا ان نقارن بأنفسنا بين ال Labels التي تنبئ بها النموذج والتي خزناها في y_pred وبين ال Labels الحقيقية والمخزونة في y_test ولكن لحسن الحظ توجد وظيفة في sklearn اسمها accuracy_score تقيس دقة هذا النموذج في التنبؤ من خلال المقارنة بين ال Labels الحقيقي للبيانات التجريبية y_test وبين ال Labels التي تنبئ بها النموذج y_pred


</p>




In [30]:
accuracy_score(y_test, y_pred)

0.78873239436619713

<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>

وكما هو واضح ان نسبة دقة هذا النموذج تقريبا 79% ولابأس بهذه النتيجة بالنسبة لأول محاولة فهي جيدة وسوف نتعلم في الدورات القادمة كيف نزيد من دقة تنبؤ النموذج من خلال عدة امور مثلا تغيير قيم المعلمات للخوارزمية وهندسة البيانات من خلال صناعة اعمدة جديدة او حذف اعمدة تضر النموذج والخ.. من العمليات 

</p>

<br>

<p style="font-family: Arial;font-size:5em; color:#09bef3; font-style:bold; text-align:center">
Congratulations
</p>
<br>
<p style="font-family: Arial; font-size:1.2em;color:black; font-style:normal; text-align:right; direction: rtl"><br>
تهانينا لقد قمت بعمل عظيم حيث تعلمت كيف تقوم ببناء تطبيق يتنبئ برطوبة الطقس اوتوماتيكيا من خلال التعلم من البيانات السابقة حيث هذا التطبيق بأمكانه التنبؤ بالرطوبة للمستقبل هل ستكون عالية ام ستكون منخفضة وهذا مايسمى بالذكاء الاصطناعي كما انه رغم ان هذا التطبيق بسيط لكنه مهم ويستخدم كثيرا وبأمكانك تضمينه في مشروعك او في موقعك الالكتروني او في بحثك او اي شيء تحتاجه كما بأمكانك خزنه وجعله كمصدر لك ترجع له في المستقبل عندما تعمل على مهام ومشاكل مشابهة لهذا المشروع بالاضافة الى انه عليك اكمال التمارين في الدورة والتي تخص هذا المشروع من اجل التدرب على بناء وتجريب النموذج بنفسك واخيرا نتمنى ان يكون قد اعجبك هذا المشروع وان تكون قد استمتعت وتعلمت الكثير في هذا المشروع

</p>

###### © Copyright LimFinity Institute of Technology 